components/patient-measurements/migrations/src/main/java/org/phenotips/measurements/internal/R71291PhenoTips1969DataMigration.java

```diff
@@ -0,0 +1,270 @@
+        EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+
+    @Inject
+    private Logger logger;
+
+    /** Resolves unprefixed document names to the current wiki. */
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> resolver;
+
+    /** Serializes the class name without the wiki prefix, to be used in the database query. */
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> serializer;
+
+    /** Resolves class names to the current wiki. */
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<EntityReference> entityResolver;
+
+    @Override
+    public String getDescription()
+    {
+        return "Migrate existing measurements to the new format";
+    }
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(71291);
+    }
+
+    @Override
+    public void hibernateMigrate() throws DataMigrationException, XWikiException
+    {
+        getStore().executeWrite(getXWikiContext(), this);
+    }
+
+    @Override
+    public Object doInHibernate(Session session) throws HibernateException, XWikiException
+    {
+        XWikiContext context = getXWikiContext();
+        XWiki xwiki = context.getWiki();
+        DocumentReference patientClassReference =
+            R71291PhenoTips1969DataMigration.this.entityResolver.resolve(PATIENT_CLASS);
+        DocumentReference oldClassReference =
+            R71291PhenoTips1969DataMigration.this.entityResolver.resolve(MEASUREMENTS_CLASS);
+        DocumentReference newClassReference =
+            R71291PhenoTips1969DataMigration.this.entityResolver.resolve(MEASUREMENT_CLASS);
+
+        Query q =
+            session.createQuery("select distinct o.name from BaseObject o where o.className = '"
+                + R71291PhenoTips1969DataMigration.this.serializer.serialize(oldClassReference) + "'");
+
+        @SuppressWarnings("unchecked")
+        List<String> documents = q.list();
+
+        for (String docName : documents) {
+            XWikiDocument doc =
+                xwiki.getDocument(R71291PhenoTips1969DataMigration.this.resolver.resolve(docName), context);
+            List<BaseObject> oldObjects = doc.getXObjects(oldClassReference);
+            if (oldObjects.isEmpty()) {
+                continue;
+            }
+
+            BaseObject patientObject = doc.getXObject(patientClassReference);
+            Date birth = null;
+            if (patientObject != null) {
+                birth = patientObject.getDateValue("date_of_birth");
+            }
+
+            for (BaseObject object : oldObjects) {
+                Date date = object.getDateValue(DATE_FIELD);
+                String age = calculateAge(birth, date);
+                Collection fieldList = object.getFieldList();
+                migrateObject(date, age, fieldList, doc, context, newClassReference);
```

Comment:
This is fine for this migration, but curious is there a better way to do migrations, without a function with side-effects? E.g. here a BaseObjects/XObject is created within the function, and is attached to the doc on creation. Seems like a better way is to return created detached XObject and attach it to the doc here, so that it is clear from reading only the high-level code what gets attached